File: backend/src/notifications/notifications.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { IsNull, Repository } from 'typeorm';
import { NotificationChannel } from '../common/entities/notification-channel.entity';
import { EncryptionService } from '../common/services/encryption.service';
import { CreateNotificationChannelDto } from './dto/create-notification-channel.dto';
import { UpdateNotificationChannelDto } from './dto/update-notification-channel.dto';
import nodemailer from 'nodemailer';

export type UsageAlertPayload = {
  tenantId: string;
  type: string;
  severity: 'warning' | 'critical';
  message: string;
  value?: number;
  limit?: number;
};

@Injectable()
export class NotificationsService {
  private transporter?: nodemailer.Transporter;

  constructor(
    @InjectRepository(NotificationChannel)
    private readonly channelsRepository: Repository<NotificationChannel>,
    private readonly encryptionService: EncryptionService
  ) {}

  async list() {
    const channels = await this.channelsRepository.find({ order: { createdAt: 'DESC' } });
    return channels.map((channel) => this.sanitize(channel));
  }

  async create(dto: CreateNotificationChannelDto) {
    const channel = this.channelsRepository.create({
      tenantId: dto.tenantId ?? null,
      type: dto.type,
      config: {
        name: dto.name ?? dto.type.toUpperCase(),
        recipients: dto.recipients ?? []
      },
      encryptedSecret: dto.webhookUrl
        ? this.encryptionService.encrypt(dto.webhookUrl)
        : null,
      enabled: dto.enabled ?? true
    });

    const saved = await this.channelsRepository.save(channel);
    return this.sanitize(saved);
  }

  async update(id: string, dto: UpdateNotificationChannelDto) {
    const channel = await this.channelsRepository.findOne({ where: { id } });
    if (!channel) {
      throw new NotFoundException('Notification channel not found');
    }

    channel.tenantId = dto.tenantId ?? channel.tenantId;
    channel.type = (dto.type as any) ?? channel.type;
    channel.enabled = dto.enabled ?? channel.enabled;

    channel.config = {
      ...(channel.config || {}),
      name: dto.name ?? (channel.config?.name as string) ?? channel.type.toUpperCase(),
      recipients: dto.recipients ?? (channel.config?.recipients as string[]) ?? []
    };

    if (dto.webhookUrl !== undefined) {
      channel.encryptedSecret = dto.webhookUrl
        ? this.encryptionService.encrypt(dto.webhookUrl)
        : null;
    }

    const saved = await this.channelsRepository.save(channel);
    return this.sanitize(saved);
  }

  async sendAlerts(tenantId: string, alerts: UsageAlertPayload[]) {
    if (alerts.length === 0) {
      return { sent: 0 };
    }

    const channels = await this.channelsRepository.find({
      where: [{ tenantId }, { tenantId: IsNull() }]
    });

    const enabled = channels.filter((channel) => channel.enabled);
    const message = this.buildMessage(tenantId, alerts);

    let sent = 0;
    for (const channel of enabled) {
      if (channel.type === 'email') {
        const recipients = (channel.config?.recipients as string[]) || [];
        const ok = await this.sendEmail(recipients, message);
        if (ok) {
          sent += 1;
        }
      }
      if (channel.type === 'slack') {
        const webhookUrl = channel.encryptedSecret
          ? this.encryptionService.decrypt(channel.encryptedSecret)
          : '';
        const ok = await this.sendSlack(webhookUrl, message);
        if (ok) {
          sent += 1;
        }
      }
    }

    return { sent };
  }

  private sanitize(channel: NotificationChannel) {
    return {
      id: channel.id,
      tenantId: channel.tenantId,
      type: channel.type,
      config: channel.config,
      enabled: channel.enabled,
      createdAt: channel.createdAt,
      updatedAt: channel.updatedAt
    };
  }

  private buildMessage(tenantId: string, alerts: UsageAlertPayload[]) {
    const lines = alerts.map((alert) => {
      const value = alert.value !== undefined && alert.limit !== undefined
        ? ` (${alert.value}/${alert.limit})`
        : '';
      return `• [${alert.severity.toUpperCase()}] ${alert.message}${value}`;
    });
    return `Alertas Provider Manager IA\nTenant: ${tenantId}\n${lines.join('\n')}`;
  }

  private getTransporter() {
    if (this.transporter) {
      return this.transporter;
    }

    const host = process.env.SMTP_HOST;
    if (!host) {
      return null;
    }

    this.transporter = nodemailer.createTransport({
      host,
      port: Number(process.env.SMTP_PORT || 587),
      secure: process.env.SMTP_SECURE === 'true',
      auth: process.env.SMTP_USER
        ? {
            user: process.env.SMTP_USER,
            pass: process.env.SMTP_PASS
          }
        : undefined
    });

    return this.transporter;
  }

  private async sendEmail(recipients: string[], message: string) {
    if (recipients.length === 0) {
      return false;
    }
    const transporter = this.getTransporter();
    if (!transporter) {
      return false;
    }

    await transporter.sendMail({
      from: process.env.SMTP_FROM || 'no-reply@provider-manager.local',
      to: recipients.join(','),
      subject: 'Alertas Provider Manager IA',
      text: message
    });

    return true;
  }

  private async sendSlack(webhookUrl: string, message: string) {
    if (!webhookUrl) {
      return false;
    }

    const response = await fetch(webhookUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ text: message })
    });

    return response.ok;
  }
}

File: backend-java/src/main/java/com/neria/manager/notifications/NotificationsService.java
package com.neria.manager.notifications;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.neria.manager.common.entities.NotificationChannel;
import com.neria.manager.common.repos.NotificationChannelRepository;
import com.neria.manager.common.services.EmailService;
import com.neria.manager.common.services.EncryptionService;
import com.neria.manager.usage.UsageAlertsService;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

@Service
public class NotificationsService {
  private final NotificationChannelRepository repository;
  private final EncryptionService encryptionService;
  private final EmailService emailService;
  private final ObjectMapper objectMapper;
  private final HttpClient httpClient = HttpClient.newHttpClient();

  public NotificationsService(
      NotificationChannelRepository repository,
      EncryptionService encryptionService,
      EmailService emailService,
      ObjectMapper objectMapper) {
    this.repository = repository;
    this.encryptionService = encryptionService;
    this.emailService = emailService;
    this.objectMapper = objectMapper;
  }

  public List<Map<String, Object>> list() {
    return repository.findAllByOrderByCreatedAtDesc().stream().map(this::sanitize).toList();
  }

  public Map<String, Object> create(CreateChannelRequest dto) {
    NotificationChannel channel = new NotificationChannel();
    channel.setId(UUID.randomUUID().toString());
    channel.setTenantId(dto.tenantId);
    channel.setType(dto.type);
    channel.setEnabled(dto.enabled != null ? dto.enabled : true);
    channel.setEncryptedSecret(
        dto.webhookUrl != null && !dto.webhookUrl.isBlank()
            ? encryptionService.encrypt(dto.webhookUrl)
            : null);
    channel.setConfig(
        toJson(
            Map.of(
                "name",
                dto.name != null ? dto.name : dto.type.toUpperCase(),
                "recipients",
                dto.recipients != null ? dto.recipients : List.of())));
    channel.setCreatedAt(LocalDateTime.now());
    channel.setUpdatedAt(LocalDateTime.now());
    return sanitize(repository.save(channel));
  }

  public Map<String, Object> update(String id, UpdateChannelRequest dto) {
    NotificationChannel channel =
        repository
            .findById(id)
            .orElseThrow(
                () ->
                    new ResponseStatusException(
                        HttpStatus.NOT_FOUND, "Notification channel not found"));
    if (dto.tenantId != null) {
      channel.setTenantId(dto.tenantId);
    }
    if (dto.type != null) {
      channel.setType(dto.type);
    }
    if (dto.enabled != null) {
      channel.setEnabled(dto.enabled);
    }
    Map<String, Object> config = parseConfig(channel.getConfig());
    if (dto.name != null) {
      config.put("name", dto.name);
    }
    if (dto.recipients != null) {
      config.put("recipients", dto.recipients);
    }
    channel.setConfig(toJson(config));
    if (dto.webhookUrl != null) {
      channel.setEncryptedSecret(
          dto.webhookUrl.isBlank() ? null : encryptionService.encrypt(dto.webhookUrl));
    }
    channel.setUpdatedAt(LocalDateTime.now());
    return sanitize(repository.save(channel));
  }

  public Map<String, Object> sendAlerts(
      String tenantId, List<UsageAlertsService.UsageAlert> alerts) {
    if (alerts.isEmpty()) {
      return Map.of("sent", 0);
    }
    List<NotificationChannel> channels =
        repository.findByTenantIdOrTenantIdIsNull(tenantId);
    List<NotificationChannel> enabled =
        channels.stream().filter(NotificationChannel::isEnabled).toList();

    String message = buildMessage(tenantId, alerts);
    int sent = 0;
    for (NotificationChannel channel : enabled) {
      if ("email".equalsIgnoreCase(channel.getType())) {
        List<String> recipients =
            (List<String>) parseConfig(channel.getConfig()).getOrDefault("recipients", List.of());
        if (!recipients.isEmpty()) {
          emailService.sendGeneric(recipients, "Alertas Provider Manager IA", message);
          sent += 1;
        }
      } else if ("slack".equalsIgnoreCase(channel.getType())) {
        String webhookUrl =
            channel.getEncryptedSecret() != null
                ? encryptionService.decrypt(channel.getEncryptedSecret())
                : null;
        if (webhookUrl != null && !webhookUrl.isBlank()) {
          if (sendSlack(webhookUrl, message)) {
            sent += 1;
          }
        }
      }
    }
    return Map.of("sent", sent);
  }

  private boolean sendSlack(String webhookUrl, String message) {
    try {
      String payload = toJson(Map.of("text", message));
      HttpRequest request =
          HttpRequest.newBuilder()
              .uri(URI.create(webhookUrl))
              .header("Content-Type", "application/json")
              .POST(HttpRequest.BodyPublishers.ofString(payload))
              .build();
      HttpResponse<String> response =
          httpClient.send(request, HttpResponse.BodyHandlers.ofString());
      return response.statusCode() >= 200 && response.statusCode() < 300;
    } catch (Exception ex) {
      return false;
    }
  }

  private Map<String, Object> sanitize(NotificationChannel channel) {
    return Map.of(
        "id",
        channel.getId(),
        "tenantId",
        channel.getTenantId(),
        "type",
        channel.getType(),
        "config",
        parseConfig(channel.getConfig()),
        "enabled",
        channel.isEnabled(),
        "createdAt",
        channel.getCreatedAt(),
        "updatedAt",
        channel.getUpdatedAt());
  }

  private String buildMessage(String tenantId, List<UsageAlertsService.UsageAlert> alerts) {
    List<String> lines = new ArrayList<>();
    for (UsageAlertsService.UsageAlert alert : alerts) {
      String value =
          alert.value() != null && alert.limit() != null
              ? " (" + alert.value() + "/" + alert.limit() + ")"
              : "";
      lines.add("• [" + alert.severity().toUpperCase() + "] " + alert.message() + value);
    }
    return "Alertas Provider Manager IA\nTenant: " + tenantId + "\n" + String.join("\n", lines);
  }

  private Map<String, Object> parseConfig(String value) {
    if (value == null || value.isBlank()) {
      return new java.util.HashMap<>();
    }
    try {
      return objectMapper.readValue(value, Map.class);
    } catch (Exception ex) {
      return new java.util.HashMap<>();
    }
  }

  private String toJson(Object payload) {
    try {
      return objectMapper.writeValueAsString(payload);
    } catch (JsonProcessingException ex) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invalid JSON payload");
    }
  }

  public static class CreateChannelRequest {
    public String tenantId;
    public String type;
    public String name;
    public List<String> recipients;
    public String webhookUrl;
    public Boolean enabled;
  }

  public static class UpdateChannelRequest {
    public String tenantId;
    public String type;
    public String name;
    public List<String> recipients;
    public String webhookUrl;
    public Boolean enabled;
  }
}
